pkg/repository/remote/upload.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"runtime"
+	"strings"
+	"time"
+
+	cjson "github.com/gibson042/canonicaljson-go"
+	"github.com/pingcap-incubator/tiup/pkg/repository/v1manifest"
+	"github.com/pingcap-incubator/tiup/pkg/utils"
+	"github.com/pingcap-incubator/tiup/pkg/version"
+)
+
+// Transporter defines methods to upload components
+type Transporter interface {
+	WithOS(os string) Transporter
+	WithArch(arch string) Transporter
+	WithDesc(desc string) Transporter
+	Open(tarbal string) error
+	Close() error
+	Upload() error
+	Sign(key *v1manifest.KeyInfo, m *v1manifest.Component) error
+}
+
+type transporter struct {
+	tarFile     *os.File
+	tarInfo     os.FileInfo
+	sha256      string
+	os          string
+	arch        string
+	entry       string
+	component   string
+	version     string
+	description string
+	endpoint    string
+}
+
+// New returns a Transporter
+func New(endpoint, component, version, entry string) Transporter {
+	return &transporter{
+		endpoint:  endpoint,
+		component: component,
+		entry:     entry,
+		os:        runtime.GOOS,
+		arch:      runtime.GOARCH,
+		version:   version,
+	}
+}
+
+func (t *transporter) WithOS(os string) Transporter {
+	t.os = os
+	return t
+}
+
+func (t *transporter) WithDesc(desc string) Transporter {
+	t.description = desc
+	return t
+}
+
+func (t *transporter) WithArch(arch string) Transporter {
+	t.arch = arch
+	return t
+}
+
+func (t *transporter) Open(tarbal string) error {
+	info, err := os.Stat(tarbal)
+	if err != nil {
+		return err
+	}
+	t.tarInfo = info
+
+	file, err := os.Open(tarbal)
+	if err != nil {
+		return err
+	}
+	t.tarFile = file
+
+	sha256, err := utils.SHA256(file)
+	if err != nil {
+		file.Close()
+		return err
+	}
+	t.sha256 = sha256
+
+	if _, err := file.Seek(0, io.SeekStart); err != nil {
+		file.Close()
+		return err
+	}
+
+	return nil
+}
+
+func (t *transporter) Close() error {
+	return t.tarFile.Close()
+}
+
+func (t *transporter) Upload() error {
+	postAddr := fmt.Sprintf("%s/api/v1/tarball/%s", t.endpoint, t.sha256)
+	tarbalName := fmt.Sprintf("%s-%s-%s-%s.tar.gz", t.component, t.version, t.os, t.arch)
```

Comment:
What if the user uploaded a file not in `tar.gz` format? I think this a potential vulnerability point, but it's ok if we don't validate it in this PR. Just remember to do that in the future.

Reply:
Yes, we should validate it in the future